sql: Support tuple column access and tuple stars

knz opened the pull request:
First commits from #26621.
Completes the fix to #24866 by re-activating disabled tests.
This work is yet another step towards #16971. It would actually fix #16971 if it were not for #26627, #26624 and #26629.

This work is yet another step towards #16971.

The labeled tuples introduced in #25283 can now be accessed using
their labels or using a star, e.g. `(E).*`.

Release note (sql change): Labeled tuples can now be accessed using
their labels (e.g. `SELECT (x).word FROM (SELECT pg_expand_keywords()
AS x)` or a star (e.g. `SELECT (x).* FROM (SELECT pg_expand_keywords()
AS x)`).

Fixes #26720.

BramGruneir committed:
sql: Support tuple column access and tuple stars

This work is yet another step towards #16971.

The labeled tuples introduced in #25283 can now be accessed using
their labels or using a star, e.g. `(E).*`.

Release note (sql change): Labeled tuples can now be accessed using
their labels (e.g. `SELECT (x).word FROM (SELECT pg_expand_keywords()
AS x)` or a star (e.g. `SELECT (x).* FROM (SELECT pg_expand_keywords()
AS x)`).

Co-authored-by: Raphael 'kena' Poss <knz@cockroachlabs.com>

BramGruneir commented:
This is looking really good!  I added a few remarks.  Also, for the optimizer stuff, it would be good to get someone from the optimizer team to take a look.

---

Reviewed 1 of 1 files at r1, 17 of 17 files at r2, 7 of 7 files at r3.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt_decompose.go, line 1519 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp1DyR43_64yaAFcKy:-LEp1DyR43_64yaAFcKz:browct7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba302d08287b0700115708f58aed44fb06c675e0/pkg/sql/opt_decompose.go#L1519)):*
> ```Go
> }
> 
> func mergeSorted(
> ```

Please tell me there is a unit test for this.  Cause it surely needs one badly.

I'm not a fan of the slicing up of the typs here.  Why not just keep two index variables around instead of this fanciness?

---

*[pkg/sql/opt_decompose.go, line 1577 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp1IW17SEToKrM9mPC:-LEp1IW17SEToKrM9mPD:b8g75ix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba302d08287b0700115708f58aed44fb06c675e0/pkg/sql/opt_decompose.go#L1577)):*
> ```Go
> }
> 
> func intersectSorted(
> ```

This too, needs a unit test too.

---

*[pkg/sql/logictest/testdata/logic_test/srfs, line 782 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp8N_tEs-jiIJwaTs5:-LEp8N_tEs-jiIJwaTs6:bwhauqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/logictest/testdata/logic_test/srfs#L782)):*
> ```
> 
> statement ok
> CREATE TABLE j(x INT PRIMARY KEY, y JSON);
> ```

Is this needed?

---

*[pkg/sql/logictest/testdata/logic_test/srfs, line 795 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp8QMu921UyNaozJqa:-LEp8QMu921UyNaozJqb:b-yxfytg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/logictest/testdata/logic_test/srfs#L795)):*
> ```
> 2  222
> 
> query I colnames
> ```

Can you add subtests for these new tests?

---

*[pkg/sql/logictest/testdata/logic_test/srfs, line 838 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp8YWZ0nVpuhWvxnzg:-LEp8YWZ0nVpuhWvxnzh:bjft926) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/logictest/testdata/logic_test/srfs#L838)):*
> ```
>     JOIN pg_catalog.pg_namespace n ON (ct.relnamespace = n.oid)
>     JOIN (SELECT i.indexrelid,
> 		 i.indrelid,
> ```

Leftover tabs

---

*[pkg/sql/logictest/testdata/logic_test/srfs, line 895 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp8ekM4p7vEd6YDmU8:-LEp8ekM4p7vEd6YDmU9:b6zi7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/logictest/testdata/logic_test/srfs#L895)):*
> ```
> # WHERE TRUE
> #   AND n.nspname = 'public'
> #   AND ct.relname = 'j'
> ```

Was this change intentional?

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 763 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp97Aa-LjJxL7qzqw7:-LEp97Aa-LjJxL7qzqw8:b-95ftrb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/logictest/testdata/logic_test/tuple#L763)):*
> ```
> 1
> 
> statement ok
> ```

Move this declaration of t down to where it is used.

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 768 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp8qv-C3AxwBOun6mw:-LEp8qv-C3AxwBOun6mx:bvxay2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/logictest/testdata/logic_test/tuple#L768)):*
> ```
> statement ok
> INSERT INTO t VALUES (1, 'one'), (2, 'two')
> ```

extra empty line

---

*[pkg/sql/opt/exec/execbuilder/scalar_builder.go, line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEozRR98Uyjt5U1twCo:-LEozRRA5q-APZH1R8I4:b-crwwkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba302d08287b0700115708f58aed44fb06c675e0/pkg/sql/opt/exec/execbuilder/scalar_builder.go#L125)):*
> ```Go
> 		}
> 	}
> 	typ := ev.Logical().Scalar.Type.(types.TTuple)
> ```

Move this into the return statement directly?

---

*[pkg/sql/opt/memo/extract.go, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEozaPj4z0YkyWFsR3-:-LEozaPj4z0YkyWFsR30:b-u48qv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba302d08287b0700115708f58aed44fb06c675e0/pkg/sql/opt/memo/extract.go#L48)):*
> ```Go
> 			datums[i] = ExtractConstDatum(ev.Child(i))
> 		}
> 		typ := ev.Logical().Scalar.Type.(types.TTuple)
> ```

This one can go in the return statement too.

---

*[pkg/sql/sem/tree/datum.go, line 2484 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp-8BCBc_wI7Zgsv0H:-LEp-8BCBc_wI7Zgsv0I:b-3zy07f) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba302d08287b0700115708f58aed44fb06c675e0/pkg/sql/sem/tree/datum.go#L2484)):*
> ```Go
> 	// The Labels sub-field can be left nil. If populated, it must have
> 	// the same arity as D.
> 	typ types.TTuple
> ```

Should this be a pointer?  Is this not adding a ttuple to every dtuple?

Also, why call this typ?  How about ttuple so there is no ambiguity?

---

*[pkg/sql/sem/tree/eval.go, line 3486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEoacPzAzykhnFjVVkb:-LEoacPzAzykhnFjVVkc:b-6juyd2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6c824e9b3b5fc1f79b483ee9d1eb2dfa54a22d/pkg/sql/sem/tree/eval.go#L3486)):*
> ```Go
> 
> // evalArgs evaluates just the function application's arguments.
> // A 1st result bool true indicates NULL should be propagated.
> ```

This is worded awkwardly.

The returned bool indicates that the NULL should be propagated.

---

*[pkg/sql/sem/tree/expr.go, line 1509 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp9eWI9iDym-hBS63j:-LEp9eWI9iDym-hBS63k:b3iilsh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/sem/tree/expr.go#L1509)):*
> ```Go
> 	Expr     Expr
> 	ColName  string
> 	Star     bool
> ```

Add a note that Star may be mutated during type check.

---

*[pkg/sql/sem/tree/type_check_test.go, line 257 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEpAUPR92Rr5RbyAnHt:-LEpAUPR92Rr5RbyAnHu:bos7gnh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/sem/tree/type_check_test.go#L257)):*
> ```Go
> 			`star expansion of tupled non-relational scalars is not supported`,
> 		},
> 		// TODO(bram): same label failure case here ********
> ```

Probably a good idea to do as my leftover todo says.  :)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-:-LEpAvDj5OvjlgcXHA7Y:b-j7ursc)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt_decompose.go, line 1519 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp1DyR43_64yaAFcKy:-LEsYXE2Dr9gMhv2F6K7:bye5fyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba302d08287b0700115708f58aed44fb06c675e0/pkg/sql/opt_decompose.go#L1519)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Please tell me there is a unit test for this.  Cause it surely needs one badly.

I'm not a fan of the slicing up of the typs here.  Why not just keep two index variables around instead of this fanciness?
</blockquote></details>

The tests are in `opt_decompose_test.go`.

But I reverted these changes. They are not needed because labels in IN comparisons never matter.

---

*[pkg/sql/opt_decompose.go, line 1577 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp1IW17SEToKrM9mPC:-LEsYc1w3SMLiYLgrD5r:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba302d08287b0700115708f58aed44fb06c675e0/pkg/sql/opt_decompose.go#L1577)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This too, needs a unit test too.
</blockquote></details>

ditto

---

*[pkg/sql/opt/exec/execbuilder/scalar_builder.go, line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEozRR98Uyjt5U1twCo:-LEsSr07CjS8EQ5v6Ess:bqzed01) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba302d08287b0700115708f58aed44fb06c675e0/pkg/sql/opt/exec/execbuilder/scalar_builder.go#L125)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Move this into the return statement directly?
</blockquote></details>

I find it harder to read.

---

*[pkg/sql/opt/memo/extract.go, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEozaPj4z0YkyWFsR3-:-LEsStd7AVW4VDcXyNUS:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba302d08287b0700115708f58aed44fb06c675e0/pkg/sql/opt/memo/extract.go#L48)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This one can go in the return statement too.
</blockquote></details>

ditto

---

*[pkg/sql/sem/tree/datum.go, line 2484 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp-8BCBc_wI7Zgsv0H:-LEsT_SmDmnB1ozZ_Wny:bxbztn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba302d08287b0700115708f58aed44fb06c675e0/pkg/sql/sem/tree/datum.go#L2484)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Should this be a pointer?  Is this not adding a ttuple to every dtuple?

Also, why call this typ?  How about ttuple so there is no ambiguity?
</blockquote></details>

The name `typ` repeats the name used throughout type_check.go (`typeAnnotation`).

The use by value instead of pointer is because of #26680. `tree.Tuple` does the same. Added a TODO.

---

*[pkg/sql/sem/tree/eval.go, line 3486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEoacPzAzykhnFjVVkb:-LEsTqwz45ZiAB7xWHYQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6c824e9b3b5fc1f79b483ee9d1eb2dfa54a22d/pkg/sql/sem/tree/eval.go#L3486)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This is worded awkwardly.

The returned bool indicates that the NULL should be propagated.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26628)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
*[pkg/sql/opt_decompose.go, line 1519 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp1DyR43_64yaAFcKy:-LEtYcMBBUhv05SUqCz0:b7lqkf9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba302d08287b0700115708f58aed44fb06c675e0/pkg/sql/opt_decompose.go#L1519)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The tests are in `opt_decompose_test.go`.

But I reverted these changes. They are not needed because labels in IN comparisons never matter.
</blockquote></details>

Ah, cool.  But if you did keep it in, you needed to augment those tests with labels.  

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26628)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
*[pkg/sql/opt/exec/execbuilder/scalar_builder.go, line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEozRR98Uyjt5U1twCo:-LEtYsVg4d9YlCm9NMCu:b32wtd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba302d08287b0700115708f58aed44fb06c675e0/pkg/sql/opt/exec/execbuilder/scalar_builder.go#L125)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I find it harder to read.
</blockquote></details>

ok

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26628)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Reviewed 22 of 22 files at r4, 17 of 17 files at r5, 7 of 7 files at r6.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26628)*
<!-- Sent from Reviewable.io -->

rytaft commented:
opt stuff LGTM

---

Reviewed 7 of 17 files at r5.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-:-LEtdkTd-rzGk1dKUd9p:bl4wfqq)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/logictest/testdata/logic_test/srfs, line 782 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp8N_tEs-jiIJwaTs5:-LEtUcgZDn4769CoCajE:b8xn9op) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/logictest/testdata/logic_test/srfs#L782)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Is this needed?
</blockquote></details>

What does the word "this" refer to in your question?

---

*[pkg/sql/logictest/testdata/logic_test/srfs, line 795 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp8QMu921UyNaozJqa:-LEtW7za6pRyyxe17pPN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/logictest/testdata/logic_test/srfs#L795)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Can you add subtests for these new tests?
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/srfs, line 838 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp8YWZ0nVpuhWvxnzg:-LEtXRGU-IlITKq0Hloh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/logictest/testdata/logic_test/srfs#L838)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Leftover tabs
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/srfs, line 895 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp8ekM4p7vEd6YDmU8:-LEtW_78EeJP9ByWQl7r:b-pyiviz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/logictest/testdata/logic_test/srfs#L895)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Was this change intentional?
</blockquote></details>

Yes because there is no index in the virtual table information_schema.columns.

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 763 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp97Aa-LjJxL7qzqw7:-LEtXMFT5oB2qFFO9y3N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/logictest/testdata/logic_test/tuple#L763)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Move this declaration of t down to where it is used.
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 768 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp8qv-C3AxwBOun6mw:-LEtXOz20cRBqF32cE8C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/logictest/testdata/logic_test/tuple#L768)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

extra empty line
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/expr.go, line 1509 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp9eWI9iDym-hBS63j:-LEuLcaY9omTpmre5qUX:b-wu3p3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/sem/tree/expr.go#L1509)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Add a note that Star may be mutated during type check.
</blockquote></details>

Changed to a separate type TupleStar.

---

*[pkg/sql/sem/tree/type_check_test.go, line 257 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEpAUPR92Rr5RbyAnHt:-LEuNTjO-LbZmmuBV29J:bq900ba) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/sem/tree/type_check_test.go#L257)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Probably a good idea to do as my leftover todo says.  :)
</blockquote></details>

Can you detail what you mean?

---

*[pkg/sql/opt_decompose.go, line 1519 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEp1DyR43_64yaAFcKy:-LEuLkNpEoiuN_hhJUzO:btd51a6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba302d08287b0700115708f58aed44fb06c675e0/pkg/sql/opt_decompose.go#L1519)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Ah, cool.  But if you did keep it in, you needed to augment those tests with labels.  
</blockquote></details>

I agree.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26628)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/sem/tree/type_check_test.go, line 257 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEpAUPR92Rr5RbyAnHt:-LEuQlIB2xDpwOM_rXcM:b-gq1vz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f87b42335c74264b074a537217fe1924c01cb714/pkg/sql/sem/tree/type_check_test.go#L257)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you detail what you mean?
</blockquote></details>

Removed comment as per offline discussion.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26628)*
<!-- Sent from Reviewable.io -->

knz commented:
@rytaft: I have extended/reworked the opt star expansion. PTAL.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/targets.go, line 70 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEuYNlj8KHqTKVCTb1w:-LEubIOUCF-LjFu_-w4J:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe422b689e2366cdde5644361004c074483642b/pkg/sql/targets.go#L70)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This is even worse than going over the line limit.  How about 

```
if hasStar, cols, typedExprs, err := sqlbase.CheckRenderStar(
  ctx, p.analyzeExpr, target, info, ivarHelper,
); err != nil {
...
```
</blockquote></details>

Good idea. Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-:-LEubOPGAjgi562MocQu:brwuie3)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Reviewed 3 of 3 files at r9.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26628)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 20 files at r7, 1 of 3 files at r9.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

---

*[pkg/sql/opt/optbuilder/project.go, line 82 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEuf1oX60dacpCZPCPA:-LEuf1oX60dacpCZPCPB:b-1bifmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/59fa89301d27495042cce005a5f0af5df85c0a78/pkg/sql/opt/optbuilder/project.go#L82)):*
> ```Go
> 		}
> 
> 		// Special handling for "*", "<table>.*" and "(E).*".
> ```

I'd change this to `(Expr).*` or `(Tuple).*` (or something along those lines) to avoid confusion

---

*[pkg/sql/opt/optbuilder/scope.go, line 533 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEufkmk1gNr6Ii1Xahj:-LEufkmk1gNr6Ii1Xahk:b-ldqj58) ([raw file](https://github.com/cockroachdb/cockroach/blob/59fa89301d27495042cce005a5f0af5df85c0a78/pkg/sql/opt/optbuilder/scope.go#L533)):*
> ```Go
> 	switch t := expr.(type) {
> 	case *tree.TupleStar:
> 		// TupleStar at the top level of a SELECT clause are replaced
> ```

TupleStar -> TupleStars (or change are -> is)

---

*[pkg/sql/opt/optbuilder/util.go, line 37 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEugDM83tVW_kuwRta5:-LEugDM9-uU-vm3XfmcU:bzbtm3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/59fa89301d27495042cce005a5f0af5df85c0a78/pkg/sql/opt/optbuilder/util.go#L37)):*
> ```Go
> 
> // expandStar expands expr into a list of columns if expr
> // corresponds to a "*" or "<table>.*".
> ```

Update comment to include TupleStar.

---

*[pkg/sql/opt/optbuilder/util.go, line 52 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEwOJ-J0WqNZY5Fddbc:-LEwOJ-KACZziS-B0-yu:b-pl4hhj) ([raw file](https://github.com/cockroachdb/cockroach/blob/59fa89301d27495042cce005a5f0af5df85c0a78/pkg/sql/opt/optbuilder/util.go#L52)):*
> ```Go
> 		// If the sub-expression is a tuple constructor, we'll de-tuplify below.
> 		// Otherwise we'll re-evaluate the expression multiple times.
> 		tTuple, isTuple := texpr.(*tree.Tuple)
> ```

When is the type `TTuple` but the expression is not actually a tuple? An example would help.

---

*[pkg/sql/opt/optbuilder/util.go, line 98 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-LEuhQSE7ou3RRaFs4r-:-LEuhQSE7ou3RRaFs4r0:b-3fy580) ([raw file](https://github.com/cockroachdb/cockroach/blob/59fa89301d27495042cce005a5f0af5df85c0a78/pkg/sql/opt/optbuilder/util.go#L98)):*
> ```Go
> 
> // expandStarAndResolveType expands expr into a list of columns if expr
> // corresponds to a "*" or "<table>.*". Otherwise, expandStarAndResolveType
> ```

Update comment to include TupleStar

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26628#-:-LEwOftkBLomZf_UnO-0:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz committed:
sql: fix _pg_expandarray() and generate_subscripts()

All the Datums generated must be distinct in memory, because different
generated rows can be stored in memory and compared against each
other.

Release note: None
